src/batcher/solve_pair.rs

```diff
@@ -1,16 +1,438 @@
+    fn match_compare(&self, other: &Order) -> Option<OrderPairType>;
+    /// Returns whether the orders can be matched. For this the tokens need
+    /// to match, there must be a price that satisfies both orders.
+    fn attracts(&self, other: &Order) -> bool;
+    /// Returns whether this order's sell token is the other order's buy token
+    /// and vice versa.
+    fn opposite_tokens(&self, other: &Order) -> bool;
+    /// Returns whether there is a price that satisfies both orders.
+    fn have_price_overlap(&self, other: &Order) -> bool;
+}
+
+
+impl Matchable for Order {
+    fn match_compare(&self, other: &Order) -> Option<OrderPairType> {
+        if !self.attracts(other) {
+            return None;
+        }
+
+        if self.buy_amount <= other.sell_amount && self.sell_amount <= other.buy_amount {
+            Some(OrderPairType::LhsFullyFilled)
+        } else if self.buy_amount >= other.sell_amount && self.sell_amount >= other.buy_amount {
+            Some(OrderPairType::RhsFullyFilled)
+        } else {
+            Some(OrderPairType::BothFullyFilled)
+        }
+    }
+
+    fn attracts(&self, other: &Order) -> bool {
+        self.opposite_tokens(other) && self.have_price_overlap(other)
+    }
+
+    fn opposite_tokens(&self, other: &Order) -> bool {
+        self.buy_token == other.sell_token && self.sell_token == other.buy_token
+    }
+
+    fn have_price_overlap(&self, other: &Order) -> bool {
+        self.sell_amount > U256::zero()
+            && other.sell_amount > U256::zero()
+            && self.buy_amount * other.buy_amount <= self.sell_amount * other.sell_amount
+    }
+}
+
+struct Match {
+    order_pair_type: OrderPairType,
+    orders: OrderPair,
+}
+
+type OrderPair = [Order; 2];
+
+type MatchedAmounts = [U256; 2];
+
 
 pub fn solve_pair(
     sell_orders_token0: &Vec<Order>,
     sell_orders_token1: &Vec<Order>,
 ) -> Result<Solution> {
+    // Can we assume that the orders are sorted already?
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token0));
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token1));
+    // Or do we need to sort them first here?
+    //sell_orders_token0.sort();
+    //sell_orders_token1.sort();
+
+    // Get number of orders in each direction.
+    let nr_orders_token0 = sell_orders_token0.len();
+    let nr_orders_token1 = sell_orders_token1.len();
+
+    // Init vectors of sell amounts.
+    let mut sell_volumes_token0 = vec![U256::zero(); nr_orders_token0];
+    let mut sell_volumes_token1 = vec![U256::zero(); nr_orders_token1];
+
+    // Match orders with best limit prices, if possible.
+    if !(sell_orders_token0.is_empty() || sell_orders_token1.is_empty()) {
+
+        // The best orders are the last elements in their vectors.
+        let best_sell_order_token0 = &sell_orders_token0[nr_orders_token0 - 1];
+        let best_sell_order_token1 = &sell_orders_token1[nr_orders_token1 - 1];
```

Comment:
Should this be assumed, or should it be part of this function's job?

Reply:
I wasn't sure. My assumption was that orders are sorted somewhere else, so I only put some asserts at the beginning of `solve_pair()`.

Reply:
What I meant is that even if the orders come up sorted, there's no guarantees that the last elements of both vectors will be matchable, right? I'm imagining the staircases [here](https://docs.google.com/document/d/1R2xyvY0KVcLeuDP_qV9FQWOUsozK7mjh-3IO6sMXtnc/edit#heading=h.rdmscaqzw5ke) could cross and continue going up in opposite directions.

Reply:
> I'm imagining the staircases here could cross and continue going up in opposite directions.

If they cross, there's also a match, right? Only if they don't cross at all there's no match. By "best orders" I mean the ones offering the lowest limit price, i.e., wanting to receive the least amount of tokens, so if these can't be matched, then nothing can matched. This is checked in the if-statement right below.

Reply:
Ah.. I thought "best" order meant the ones with "better" price for the order owners. Everything is cool then, thanks!